alphabase/spectral_library/annotate.py

```diff
@@ -0,0 +1,557 @@
+    stop_indices = flatlib.precursor_df["frag_stop_idx"].to_numpy()
+
+    def gini(intensities: np.ndarray) -> float:
+        # Handle empty or zero cases
+        if len(intensities) == 0 or np.sum(intensities) == 0:
+            return 1.0
+
+        # Sort intensities in ascending order
+        sorted_intensities = np.sort(intensities)
+        n = len(sorted_intensities)
+        index = np.arange(1, n + 1)
+
+        # Create intermediate values to improve readability and reduce in-place operations
+        intensity_sum = np.sum(sorted_intensities)
+        weighted_sum = np.sum(index * sorted_intensities)
+
+        return (2 * weighted_sum) / (n * intensity_sum) - ((n + 1) / n)
+
+    flatlib.precursor_df["sequence_gini"] = np.array(
+        [
+            gini(flatlib._fragment_intensity_df.to_numpy()[start:stop].flatten())  # noqa: SLF001
+            for start, stop in zip(start_indices, stop_indices)
+        ]
+    )
+
+
+def _normalized_count_metric(flatlib: SpecLibFlat) -> None:
+    """Calculate the number of fragments per precursor normalized by the precursor length.
+
+    Important:
+    ---------
+    This function operates in place and modifies the flatlib object.
+
+    """
+    # Get all values at once to avoid repeated access
+    fragment_values = flatlib._fragment_intensity_df.to_numpy()  # noqa: SLF001
+    start_indices = flatlib.precursor_df["frag_start_idx"].to_numpy()
+    stop_indices = flatlib.precursor_df["frag_stop_idx"].to_numpy()
+
+    # Calculate lengths once
+    lengths = stop_indices - start_indices
+
+    # Calculate counts using vectorized operations
+    counts = np.array(
+        [
+            np.sum(fragment_values[start:stop].flatten() > 0) / length
+            for start, stop, length in zip(start_indices, stop_indices, lengths)
+        ]
+    )
+
+    flatlib.precursor_df["normalized_count"] = counts
+
+
+def _nan_median(series: np.ndarray) -> float:
+    """Calculate the median of a numpy array, ignoring NaNs and infs.
+
+    defaults to inf if there are no valid values
+    """
+    valid_values = series[~np.isnan(series) & ~np.isinf(series)]
+    if len(valid_values) == 0:
+        return np.inf
+    return np.median(valid_values)
+
+
+def _mass_accuracy_metric(flatlib: SpecLibFlat) -> None:
+    """Calculate the mass accuracy metric for the precursors.
+
+    Important:
+    ---------
+    This funmction operates in place and modifies the flatlib object
+
+    """
+    # Vectorized operation to calculate coverage for all precursors at once
+    start_indices = flatlib.precursor_df["frag_start_idx"].to_numpy()
+    stop_indices = flatlib.precursor_df["frag_stop_idx"].to_numpy()
+
+    flatlib.precursor_df["mass_accuracy"] = np.array(
+        [
+            _nan_median(flatlib._fragment_df["error"].to_numpy()[start:stop])  # noqa: SLF001
+            for start, stop in zip(start_indices, stop_indices)
+        ]
+    )
+
+
+def add_dense_lib(
```

Comment:
what do you think about adding this function directly to the speclib flat object?
so we can call `SpecLibFlat.add_dense_lib()`.
I think this fits the ab pattern better.